### DISCUSSION TITLE
Reticulum device survey

### DISCUSSION THREAD

**Discussion by jacobeva** [markqvist/Reticulum#394] [General]
I'm looking at developing a pre-built LoRa radio to sell as a product for Reticulum, I'm currently running a survey to get user input on what to prioritise in the design, please find it below:
https://forms.gle/uffymAM9ikY7hUj68
Your input is very much appreciated, thank you! :)

**Comment by kc1awv** [comment]
To be honest, I'd love to see BLE implemented. I've been hacking away at RNode firmware for a bit, transferring everything to Platformio for my own devices, but all the newer (S3) ESP32 devices I have are stuck because of Bluetooth Serial not present in the BLE stack(?). I've been taking a look at some other programs/firmware for hints on how to get this implemented, but alas day job has to take precedence :smile:

**Reply by jacobeva** [reply]
Sorry, what do you mean by BLE support, isn't this already implemented in the RNode firmware for pairing with devices like mobile phones? Or am I misunderstanding :)

**Reply by kc1awv** [reply]
BLE (Bluetooth Low Energy) != Bluetooth classic. BLE is better suited for low-power devices where constant streaming data (like audio, for music and phone calls for example) isn't needed. Some newer ESP32 devices are already implementing this, and requires a wholly different library than what is currently in RNode firmware. I've tried compiling RNode for an ESP32-S3-DevKitC-1-N16R8V, but it fails because Bluetooth classic isn't available and uses BLE instead.

**Reply by jacobeva** [reply]
Ah, I see, thanks for the clarification!

**Reply by cobraPA** [reply]
The Android RNode support uses Bluetooth classic as noted.  I am working on a BLE version for RNodes that would work with iPhones and you can read about it over in the iPhone discussion here.
Having said that, I think Bluetooth classic is still supported in the bluetooth libraries for most Arduino variants?  It's just the libraries you use.  Although maybe some versions don't have support for both anymore?
But you are right, I am writing around bluetooth serial with a direct BLE read/write implementation for now.  It's not really necessary to use bluetooth serial, I don't think.  I guess we will know soon enough!

**Reply by kc1awv** [reply]
ESP32-S3 is BT 5, which is BLE exclusive (if I read the spec correctly). So, the Bluetooth classic protocol doesn't look like it's there and the (older) library used probably won't work.

[ESP32-S3 Datasheet](https://www.espressif.com/sites/default/files/documentation/esp32-s3_datasheet_en.pdf)

![Screenshot 2023-12-02 111007](https://github.com/markqvist/Reticulum/assets/43918257/c8b754db-a2ff-4b50-b4e0-14436cf38762)

**Reply by cobraPA** [reply]
Yes, I see.  It's a lower cost SOC that doesn't support Bluetooth standard...  I'd actually avoided buying any ESP32 variants like that, since it wouldn't currently be possible to use the Android support in present RNode firmware and the Reticulum stack.
Note there may also be performance penalties as well.  BLE looks like it supports lower transfer rates in general, than bluetooth standard.  I think this was part of the reason Mark went with Bluetooth standard on the Android implementation.
On the flip side, Apple limits use of Bluetooth standard, I assume because of power usage concerns.  So it looks like BLE is the way to go on iOS/iPadOS platforms.  We'll have to see if performance is an issue or not once we can do some testing.

**Comment by cobraPA** [comment]
I saw some BLE flags show up in Mark's RNode firmware and wanted to update this topic.  I have a working (test) version of BLE on iPhone with a custom RNode firmware.  I plan to update the git version soon so I can push out a new test version of the iPhone Sideband app with RNode support.  @markqvist, maybe the BLE implementation would be useful for the support that was recently merged.  I didn't implement a shared version that supports BLE and Standard Bluetooth (yet), but I have a note in the code that NimBLE should be able to do that.  BLE only version here (to be updated shortly to my current working version): https://github.com/cobraPA/RNode_Firmware_iOS/commits/kb_ble_test1

**Comment by jacobeva** [comment]
Good job buddy!

**Reply by cobraPA** [reply]
@jacobeva Thanks!  I just found some of the discussion between you and Mark over in the Rnode firmware git when you added SX1262 and nRF.  Nice work!  I saw you were looking into BLE also... is there more work you are doing there?  Any discussion or PRs I should look at?  I have a working version for Heltec Lora32 V2 style devices / iPhone and am attempting to run one with a Heltec Lora32 V3 SX1262 now as well.  Just curious if you are also doing additional BLE work and what platforms you are trying.

**Reply by jacobeva** [reply]
Hello buddy! BLE support is all implemented on the hardware side (nRF52
only), it just must be supported on the mobile applications currently.
My BLE implementation definitely works, I've tested it using the
ble-serial tool on my laptop. However, I have not gone anywhere near
adding support to sideband. For one, I don't even know anyone who can
compile sideband besides Mark lol, and I also have no clue what I'm
doing, nor the time to bother with it at the moment.

I am not doing additional BLE work for now, my PRs were mainly to add
support for the hardware I intend to use to satisfy the consumer
requirements from my survey! However, further development for other
platforms is something I may consider in the future when it is
convenient for me. If you would like any assistance to this end, please
feel free to reach out to me (I'm in the Reticulum Matrix room), but for
now I am catching up on important work! :)

**Reply by jacobeva** [reply]
Sorry if you got notified twice! When I hit reply to email, it started a new thread! Oh github...
